test/src/test/java/jenkins/bugs/Jenkins64991Test.java

```diff
@@ -0,0 +1,163 @@
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import com.gargoylesoftware.htmlunit.html.HtmlPasswordInput;
+import com.gargoylesoftware.htmlunit.html.HtmlTextInput;
+import hudson.model.FreeStyleProject;
+import hudson.security.Messages;
+import hudson.security.Permission;
+import jenkins.model.Jenkins;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.MockAuthorizationStrategy;
+import org.xml.sax.SAXException;
+
+import java.io.IOException;
+
+import static org.hamcrest.CoreMatchers.containsStringIgnoringCase;
+import static org.hamcrest.CoreMatchers.endsWithIgnoringCase;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class Jenkins64991Test {
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Before
+    public void setUp() throws Exception {
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy().grant(Permission.READ).everywhere().toEveryone().grant(Jenkins.ADMINISTER).everywhere().to("alice"));
+    }
+
+    @Test
+    public void test403Redirect() throws Exception {
+        final JenkinsRule.WebClient webClient = j.createWebClient().withThrowExceptionOnFailingStatusCode(false);
+        final HtmlPage loginPage = webClient.goTo("manage");
+
+        assertTrue(loginPage.isHtmlPage());
+        assertThat(loginPage.getUrl().toExternalForm(), containsStringIgnoringCase("%2Fmanage"));
+
+        ((HtmlTextInput)loginPage.getElementByName("j_username")).setText("alice");
+        ((HtmlPasswordInput)loginPage.getElementByName("j_password")).setText("alice");
+
+        final Page redirectedPage = HtmlFormUtil.submit(loginPage.getFormByName("login"));
+        assertTrue(redirectedPage.isHtmlPage());
+        assertEquals(j.getURL() + "manage", redirectedPage.getUrl().toExternalForm());
+        assertThat(redirectedPage.getWebResponse().getContentAsString(), containsStringIgnoringCase(Messages.GlobalSecurityConfiguration_DisplayName()));
+    }
+
+
+    @Test
+    public void testRedirect() throws Exception {
+        final JenkinsRule.WebClient webClient = j.createWebClient();
+        final HtmlPage indexPage = webClient.goTo("");
+
+        final Page loginPage = indexPage.getElementsByTagName("a").stream().filter(
+                e -> e.hasAttribute("href") && e.getAttribute("href").contains(j.jenkins.getSecurityRealm().getLoginUrl())
+        ).findFirst().orElseThrow(() -> new RuntimeException("cannot find login link")).click();
+        // Could be simplified to `indexPage.getElementById("login-link").click();` if we're willing to edit loginLink.jelly
+
+        assertTrue(loginPage.isHtmlPage());
+        assertThat(loginPage.getUrl().toExternalForm(), endsWithIgnoringCase("%2F"));
```

Comment:
Somewhat weak assertion. Will it end with just `%2F`, or more specifically `%2Fjenkins%2F` or `?from=%2F` etc.?